### ISSUE TITLE
[Feature request] Alerting: support anomaly detection algorithms

### ISSUE THREAD

**Issue by gtt116** [grafana/grafana#8346]
Grafana Version: 4.2.0
Data source: Graphite and InfluxDB

Many time the static threshold alerting not work well. Take QPS for example, we can't setup a maximum or minimum threshold, because QPS may always changing. To monitoring these kind of seasonal metrics maybe anomaly detection algorithm can work.

A concrete example is Datadog, [It supports four anomaly detection algorithms](http://docs.datadoghq.com/guides/outliers/), which user just picks up a function that match his requirement, and tune the arguments of the function to make the algorithm works well.

To make grafana more intelligent, maybe we can support some functions like what Datadog can do?

Cheers!

**Comment by torkelo** [comment]
These algorithms are better (or some cases must be) implemented in the time series database and it's query language. 

The datadog functions you speak of already exist in some data sources Grafana support (like Graphite). So you will need to request this feature from InfluxDB.

**Comment by gtt116** [comment]
I am also thinking about where these evaluation function should stay. 

IMHO, If we implement at grafana, grafana will be more like a generic alarm engine, no matter what data sources we used, grafana still has the ability to find anomaly and fire an notification. I believe this idea is similar to what we already done in grafana: all the [evaluator/reducer](https://github.com/grafana/grafana/blob/master/pkg/services/alerting/conditions/reducer.go) algorithms are implement in grafana, which don't depend on any data source.  

The another reason that I guest grafana is the better place is that grafana has the ability to create awesome visualization, which combines anomaly detection result and real data in one dashboard/graph, like this:

![image](https://cloud.githubusercontent.com/assets/1257631/25979484/50ce55a2-36fb-11e7-9dd7-1ca58d259996.png)

and this:
![image](https://cloud.githubusercontent.com/assets/1257631/25978870/9655bbe6-36f7-11e7-99c5-34e5d3a1aa1a.png)

Does it make sense?

**Comment by torkelo** [comment]
Do that Grafana well grafana needs access to the raw data which it does not have, grafana can visualize the result of the anomaly detection functions, but it's hard for grafana to perform them without download the whole database

**Comment by EugeneKrapivin** [comment]
Could you create your own data as it comes and store it for future reference?

By create I mean analyze metrics from the data source, calculate a baseline and store it as two metrics (upper-lower) in the data source. This way creating adaptive alerting would be a matter using the upper-lower metrics as alerting boundaries (or using any other metric as an alert boundary).

Or I could be getting it totally wrong.

**Comment by jdraymon** [comment]
@EugeneKrapivin yes, that's a good solve, but dependent on #6593 - but the writing of what makes for an anomalous condition as a metric is better implemented either by TSDB or your business logic watching the TSDB

**Comment by EugeneKrapivin** [comment]
@jdraymon of course its better that "anomaly boundaries" will be calculated in the data source itself. 
However, its hard to expect all major sources to implement that logic and make it easy to integrate with Grafana. 

What I suggest is add optional interface for Grafana to calculate and write those metrics into a data source and later use it to set alerts as the "live" metric crosses the projected boundaries. 

I think the reasoning for this implementation should be quite similar to why Grafana got Alerting in the first place, most sources could have provided such functionality (and some do) out-of-the-box or as plugins.

**Comment by santi8ago8** [comment]
+1

**Comment by kjaanson** [comment]
+1 for this

Have come upon many cases where it is convenient or necessary for the anomaly detection to happen much closer on the data monitoring/visualisation side simply because not every data source supports AD. We have developed adhoc custom solutions for these cases so far but solution on Grafana together with its alerting would be much more convenient.

**Comment by fcuenca4** [comment]
+1

**Comment by erraggy** [comment]
![GIVE IT TO ME](https://media.giphy.com/media/fJW2jK0akgkhi/giphy.gif)

**Comment by colszowka** [comment]
Everyone please stop spamming with +1 or equivalent GIFs and just make your voice heard by upvoting the OP or submitting a PR. Thanks.

**Comment by veesahni** [comment]
[Cloudwatch recently added anomaly detection](https://aws.amazon.com/blogs/aws/new-amazon-cloudwatch-anomaly-detection/).  The user interface is quite simple - pick a metric, enable anomaly detection and customize parameters if you need.   Similar to what datadog implemented.

GitLab also posted [this article](https://about.gitlab.com/blog/2019/07/23/anomaly-detection-using-prometheus/) on how they implement anomaly detection in prometheus. Do see the video in there.

From a user experience perspective, it makes sense for Grafana to support some kind of "detect anomalies on this query" feature.  But what I'm reading in this thread is that the data source really needs to help out here.

So thinking about how these features are getting implemented:
* A prediction time series is being generated based on historical data
* That prediction time series is combined with allowable stddev to create two time series (an upper bound and a lower bound)

To the TSDB, this is one "hard" query (a prediction) and two simple functions running on it (stddev).

To Grafana, this is two extra time series (upper and lower bounds) that need to be visualized and and alert based on 1 time series crossing the other.

For the query, the GitLab article provides a recipe.  They implemented it using recording rules but prometheus subquery support should make it possible to implement that just as a query that Grafana could generate on the fly.

For the alert, #6593 is already tracking the underlying requirement.

**Comment by ramnes** [comment]
I think (?) it's still relevant

**Comment by EugeneKrapivin** [comment]
It is, but it really seems that it ain't gonna happen. Thinking about it,
even if it does it'll be part of the enterprise edition of grafana rather
than part of the OSS.

**Comment by jtheory** [comment]
:wave: I can add some internal context here; it's been a while since this issue first came up and Torkel added his thoughts.

Some of the anomaly detection Torkel was hoping for **is** now built into Prometheus; you can use aggregations over time like `stddev_over_time` and `mad_over_time` (experimental) to build basic no-training-needed anomaly detection ([example](https://blog.devops.dev/anomaly-alerts-for-monitoring-using-grafana-and-prometheus-8f1aaf3de54f)).

There is also now more robust anomaly detection that we've built into Grafana Cloud, in [the Machine Learning plugin](https://grafana.com/docs/grafana-cloud/alerting-and-irm/machine-learning/) -- you can configure a forecast (and define holidays where your usual seasonality will break), we'll run daily training with Prophet, and expose the results in a Prometheus-compatible datasource, for dashboards or alerting. There's also support for outlier detection using a few different algorithms (e.g., you have the same metrics on a few hundred servers; which ones are not like the others?). It's there even in a free account.

But it's currently not OSS (or even Enterprise) because the training, derived metric storage etc runs on infrastructure we've scaled up in Cloud. It'd be possible to package it up to run outside of Cloud (Enterprise and/or OSS) but we've not done that work yet, and we don't hear a lot of demand for it -- many larger on-prem Grafana installations are fine using a hybrid solution, leaning on Cloud for stuff like ML and IRM.

I'd be interested to hear an update of what's possible now with e.g. `mad_over_time` in Prometheus and new Grafana Alerting, vs. what kind of needs people still have for OSS and at smaller scale. There's other stuff we've built, including purely FE forecasting, that's still at more of a hackathon stage.
